File: recoup/python-backend/services/pricing.py
"""
Pricing Utilities & Calculations
Converted from relay/lib/pricing.ts

Pricing V3 Structure:
- Starter: £19/month or £182/year (10 collections)
- Growth: £39/month or £374/year (50 collections)
- Pro: £75/month or £720/year (unlimited)

Annual discount: 20% (equivalent to 2.4 months free)
"""

from typing import Optional, Literal, Dict
from dataclasses import dataclass

PricingTier = Literal['starter', 'growth', 'pro']
LegacyTier = Literal['free', 'paid', 'business']
AllTiers = Literal['starter', 'growth', 'pro', 'free', 'paid', 'business']


@dataclass
class TierPricing:
    """Pricing tier information"""
    monthly: float
    annual: float
    annual_savings: float
    collections_limit: Optional[int]  # None = unlimited
    team_members: Optional[int]  # None = unlimited
    features: list[str]


PRICING_TIERS: Dict[str, TierPricing] = {
    'starter': TierPricing(
        monthly=19,
        annual=182,  # £19 × 12 × 0.8 = £182.40 (rounded down)
        annual_savings=46,  # £228 - £182 = £46
        collections_limit=10,
        team_members=1,
        features=[
            'Email reminders',
            'Manual collection tracking',
            'Invoice management',
            'Payment claims',
            'Email support (48h response)',
        ],
    ),
    'growth': TierPricing(
        monthly=39,
        annual=374,  # £39 × 12 × 0.8 = £374.40 (rounded down)
        annual_savings=94,  # £468 - £374 = £94
        collections_limit=50,
        team_members=5,
        features=[
            'Smart reminders (Email + SMS + WhatsApp)',
            'Payment verification system',
            'Collections escalation automation',
            'Basic AI analytics',
            'Behavioral email sequences',
            'Email support (24h response)',
        ],
    ),
    'pro': TierPricing(
        monthly=75,
        annual=720,  # £75 × 12 × 0.8 = £720
        annual_savings=180,  # £900 - £720 = £180
        collections_limit=None,  # Unlimited
        team_members=None,  # Unlimited
        features=[
            'All channels (Email/SMS/WhatsApp/Phone)',
            'AI-powered recovery strategies',
            'Advanced analytics & insights',
            'Custom escalation workflows',
            'API access & integrations',
            'Dedicated account manager',
            'Priority support (2h response)',
        ],
    ),
}


def get_tier_price(tier: PricingTier, is_annual: bool = False) -> float:
    """
    Get price for a tier

    Args:
        tier: Tier name
        is_annual: Whether annual billing

    Returns:
        Price in GBP
    """
    pricing = PRICING_TIERS[tier]
    return pricing.annual if is_annual else pricing.monthly


def get_tier_collections_limit(tier: PricingTier) -> Optional[int]:
    """
    Get collections limit for a tier

    Args:
        tier: Tier name

    Returns:
        Collections per month (None = unlimited)
    """
    return PRICING_TIERS[tier].collections_limit


def get_tier_team_members_limit(tier: PricingTier) -> Optional[int]:
    """
    Get team members limit for a tier

    Args:
        tier: Tier name

    Returns:
        Team members allowed (None = unlimited)
    """
    return PRICING_TIERS[tier].team_members


def get_annual_savings(tier: PricingTier) -> float:
    """
    Calculate annual savings

    Args:
        tier: Tier name

    Returns:
        Savings in GBP when paying annually
    """
    return PRICING_TIERS[tier].annual_savings


def get_monthly_equivalent_price(tier: PricingTier) -> float:
    """
    Calculate monthly equivalent price for annual plans

    Args:
        tier: Tier name

    Returns:
        Monthly equivalent price (annual / 12)
    """
    return round(PRICING_TIERS[tier].annual / 12)


def map_legacy_tier_to_v3(legacy_tier: LegacyTier) -> PricingTier:
    """
    Map legacy tier to Pricing V3 tier

    Migration mapping:
    - free → starter (with trial)
    - paid → growth (legacy catch-all)
    - business → pro (upgrade path)

    Args:
        legacy_tier: Old subscription tier

    Returns:
        Equivalent V3 tier
    """
    mapping: Dict[LegacyTier, PricingTier] = {
        'free': 'starter',
        'paid': 'growth',
        'business': 'pro',
    }

    return mapping[legacy_tier]


def has_exceeded_collections_limit(tier: PricingTier, collections_used: int) -> bool:
    """
    Check if user has exceeded their collections limit

    Args:
        tier: User's subscription tier
        collections_used: Collections used this month

    Returns:
        True if limit exceeded
    """
    limit = get_tier_collections_limit(tier)

    # Unlimited tiers never exceed
    if limit is None:
        return False

    return collections_used >= limit


def calculate_overage_cost(tier: PricingTier, collections_used: int) -> float:
    """
    Calculate overage cost for additional collections

    Based on pricing framework (£1-2 per collection)

    Args:
        tier: User's subscription tier
        collections_used: Collections used this month

    Returns:
        Overage cost in GBP
    """
    limit = get_tier_collections_limit(tier)

    # No overage for unlimited tiers
    if limit is None:
        return 0

    # No overage if under limit
    if collections_used <= limit:
        return 0

    overage = collections_used - limit

    # Tiered overage pricing
    if tier == 'starter':
        return overage * 2  # £2 per collection for Starter
    elif tier == 'growth':
        return overage * 1.5  # £1.50 per collection for Growth

    return 0  # Pro has unlimited, no overage


def get_recommended_upgrade(
    current_tier: PricingTier,
    collections_used: int
) -> Optional[PricingTier]:
    """
    Get recommended upgrade tier

    Suggests upgrade if user is consistently hitting limits

    Args:
        current_tier: User's current tier
        collections_used: Collections used this month

    Returns:
        Recommended tier or None if no upgrade needed
    """
    limit = get_tier_collections_limit(current_tier)

    # No upgrade needed for Pro tier
    if current_tier == 'pro':
        return None

    # If no limit, no upgrade needed
    if limit is None:
        return None

    # Recommend upgrade if using >80% of limit
    usage_percentage = (collections_used / limit) * 100

    if usage_percentage >= 80:
        if current_tier == 'starter':
            return 'growth'
        if current_tier == 'growth':
            return 'pro'

    return None


def calculate_ltv(tier: PricingTier, is_annual: bool = False) -> float:
    """
    Calculate LTV (Lifetime Value) for a tier

    Assumptions:
    - Average retention: 12 months (industry standard for SMB SaaS)
    - Churn rate: ~8.3% per month (1/12)

    Args:
        tier: Subscription tier
        is_annual: Whether annual billing

    Returns:
        Estimated LTV in GBP
    """
    monthly_price = get_tier_price(tier, False)
    average_retention_months = 12

    if is_annual:
        # Annual customers typically have 1.5x longer retention
        return get_tier_price(tier, True) * 1.5

    return monthly_price * average_retention_months


def format_price(price: float, show_pence: bool = False) -> str:
    """
    Format price for display

    Args:
        price: Price in GBP
        show_pence: Whether to show .00 decimals

    Returns:
        Formatted price string (e.g., "£39" or "£39.00")
    """
    if show_pence or price % 1 != 0:
        return f"£{price:.2f}"
    return f"£{int(price)}"


def get_annual_discount_percentage(tier: PricingTier) -> int:
    """
    Calculate discount percentage for annual plans

    Args:
        tier: Tier name

    Returns:
        Discount percentage (e.g., 20 for 20%)
    """
    monthly = get_tier_price(tier, False)
    annual = get_tier_price(tier, True)

    full_year_cost = monthly * 12
    savings = full_year_cost - annual

    return round((savings / full_year_cost) * 100)

File: recoup/lib/pricing.ts
/**
 * Pricing Utilities & Calculations
 *
 * Based on: pricing-implementation-framework.md §1-2
 * Research: saas-pricing-optimization-guide.md
 *
 * Pricing V3 Structure:
 * - FREE: 5 collections/month (truly free, not just demo)
 * - Starter: £19/month or £182/year (10 collections)
 * - Growth: £39/month or £374/year (50 collections)
 * - Pro: £75/month or £720/year (unlimited)
 *
 * Optional Add-ons:
 * - HMRC MTD: £20/month or £200/year (VAT submissions, can be added to any tier)
 *
 * Annual discount: 20% (equivalent to 2.4 months free)
 *
 * Phase 2 Task 8
 */

export type PricingTier = 'free' | 'starter' | 'growth' | 'pro';
export type LegacyTier = 'paid' | 'business';
export type AllTiers = PricingTier | LegacyTier;
export type AddonType = 'hmrc_mtd';

export interface TierPricing {
    monthly: number;
    annual: number;
    annualSavings: number;
    collectionsLimit: number | null; // null = unlimited
    teamMembers: number | null; // null = unlimited
    features: string[];
}

export const PRICING_TIERS: Record<PricingTier, TierPricing> = {
    free: {
        monthly: 0,
        annual: 0,
        annualSavings: 0,
        collectionsLimit: 5, // Increased from 1 to 5 - truly usable free tier
        teamMembers: 1,
        features: [
            'Unlimited invoices',
            '5 collections per month',
            'Email reminders',
            'BACS "I Paid" button',
            'Manual payment tracking',
            'Community support',
        ],
    },
    starter: {
        monthly: 19,
        annual: 182, // £19 × 12 × 0.8 = £182.40 (rounded down)
        annualSavings: 46, // £228 - £182 = £46
        collectionsLimit: 10,
        teamMembers: 1,
        features: [
            'Email reminders',
            'Manual collection tracking',
            'Invoice management',
            'Payment claims',
            'Email support (48h response)',
        ],
    },
    growth: {
        monthly: 39,
        annual: 374, // £39 × 12 × 0.8 = £374.40 (rounded down)
        annualSavings: 94, // £468 - £374 = £94
        collectionsLimit: 50,
        teamMembers: 5,
        features: [
            'Smart reminders (Email + SMS + WhatsApp)',
            'Payment verification system',
            'Collections escalation automation',
            'Basic AI analytics',
            'Behavioral email sequences',
            'Email support (24h response)',
        ],
    },
    pro: {
        monthly: 75,
        annual: 720, // £75 × 12 × 0.8 = £720
        annualSavings: 180, // £900 - £720 = £180
        collectionsLimit: null, // Unlimited
        teamMembers: null, // Unlimited
        features: [
            'All channels (Email/SMS/WhatsApp/Phone)',
            'AI-powered recovery strategies',
            'Advanced analytics & insights',
            'Custom escalation workflows',
            'API access & integrations',
            'Dedicated account manager',
            'Priority support (2h response)',
        ],
    },
};

/**
 * Add-on Pricing (HMRC Making Tax Digital)
 *
 * Separate subscription that can be added to any tier
 */
export interface AddonPricing {
    monthly: number;
    annual: number;
    annualSavings: number;
    features: string[];
}

export const ADDON_PRICING: Record<AddonType, AddonPricing> = {
    hmrc_mtd: {
        monthly: 20,
        annual: 200, // £20 × 12 × 0.833 = £200 (17% discount)
        annualSavings: 40, // £240 - £200 = £40
        features: [
            'Unlimited VAT return submissions',
            'Automated obligation tracking',
            'Quarterly deadline reminders',
            'VAT calculation dashboard',
            'HMRC OAuth integration',
            'FCA-compliant audit trail',
        ],
    },
};

/**
 * Get price for a tier
 *
 * @param tier - Tier name
 * @param isAnnual - Whether annual billing
 * @returns Price in GBP
 */
export function getTierPrice(tier: PricingTier, isAnnual: boolean = false): number {
    const pricing = PRICING_TIERS[tier];
    return isAnnual ? pricing.annual : pricing.monthly;
}

/**
 * Get collections limit for a tier
 * 
 * @param tier - Tier name
 * @returns Collections per month (null = unlimited)
 */
export function getTierCollectionsLimit(tier: PricingTier): number | null {
    return PRICING_TIERS[tier].collectionsLimit;
}

/**
 * Get team members limit for a tier
 * 
 * @param tier - Tier name
 * @returns Team members allowed (null = unlimited)
 */
export function getTierTeamMembersLimit(tier: PricingTier): number | null {
    return PRICING_TIERS[tier].teamMembers;
}

/**
 * Calculate annual savings
 * 
 * @param tier - Tier name
 * @returns Savings in GBP when paying annually
 */
export function getAnnualSavings(tier: PricingTier): number {
    return PRICING_TIERS[tier].annualSavings;
}

/**
 * Calculate monthly equivalent price for annual plans
 * 
 * @param tier - Tier name
 * @returns Monthly equivalent price (annual / 12)
 */
export function getMonthlyEquivalentPrice(tier: PricingTier): number {
    return Math.round(PRICING_TIERS[tier].annual / 12);
}

/**
 * Map legacy tier to Pricing V3 tier
 * 
 * Migration mapping:
 * - free → starter (with trial)
 * - paid → growth (legacy catch-all)
 * - business → pro (upgrade path)
 * 
 * @param legacyTier - Old subscription tier
 * @returns Equivalent V3 tier
 */
export function mapLegacyTierToV3(legacyTier: LegacyTier): PricingTier {
    const mapping: Record<LegacyTier, PricingTier> = {
        free: 'starter',
        paid: 'growth',
        business: 'pro',
    };

    return mapping[legacyTier];
}

/**
 * Check if user has exceeded their collections limit
 * 
 * @param tier - User's subscription tier
 * @param collectionsUsed - Collections used this month
 * @returns true if limit exceeded
 */
export function hasExceededCollectionsLimit(
    tier: PricingTier,
    collectionsUsed: number
): boolean {
    const limit = getTierCollectionsLimit(tier);

    // Unlimited tiers never exceed
    if (limit === null) return false;

    return collectionsUsed >= limit;
}

/**
 * Calculate overage cost for additional collections
 * 
 * Based on: pricing-implementation-framework.md (£1-2 per collection)
 * 
 * @param tier - User's subscription tier
 * @param collectionsUsed - Collections used this month
 * @returns Overage cost in GBP
 */
export function calculateOverageCost(
    tier: PricingTier,
    collectionsUsed: number
): number {
    const limit = getTierCollectionsLimit(tier);

    // No overage for unlimited tiers
    if (limit === null) return 0;

    // No overage if under limit
    if (collectionsUsed <= limit) return 0;

    const overage = collectionsUsed - limit;

    // Tiered overage pricing
    if (tier === 'starter') {
        return overage * 2; // £2 per collection for Starter
    } else if (tier === 'growth') {
        return overage * 1.5; // £1.50 per collection for Growth
    }

    return 0; // Pro has unlimited, no overage
}

/**
 * Get recommended upgrade tier
 * 
 * Suggests upgrade if user is consistently hitting limits
 * 
 * @param currentTier - User's current tier
 * @param collectionsUsed - Collections used this month
 * @returns Recommended tier or null if no upgrade needed
 */
export function getRecommendedUpgrade(
    currentTier: PricingTier,
    collectionsUsed: number
): PricingTier | null {
    const limit = getTierCollectionsLimit(currentTier);

    // No upgrade needed for Pro tier
    if (currentTier === 'pro') return null;

    // If no limit, no upgrade needed
    if (limit === null) return null;

    // Recommend upgrade if using >80% of limit
    const usagePercentage = (collectionsUsed / limit) * 100;

    if (usagePercentage >= 80) {
        if (currentTier === 'starter') return 'growth';
        if (currentTier === 'growth') return 'pro';
    }

    return null;
}

/**
 * Calculate LTV (Lifetime Value) for a tier
 * 
 * Assumptions:
 * - Average retention: 12 months (industry standard for SMB SaaS)
 * - Churn rate: ~8.3% per month (1/12)
 * 
 * @param tier - Subscription tier
 * @param isAnnual - Whether annual billing
 * @returns Estimated LTV in GBP
 */
export function calculateLTV(tier: PricingTier, isAnnual: boolean = false): number {
    const monthlyPrice = getTierPrice(tier, false);
    const averageRetentionMonths = 12;

    if (isAnnual) {
        // Annual customers typically have 1.5x longer retention
        return getTierPrice(tier, true) * 1.5;
    }

    return monthlyPrice * averageRetentionMonths;
}

/**
 * Format price for display
 * 
 * @param price - Price in GBP
 * @param showPence - Whether to show .00 decimals
 * @returns Formatted price string (e.g., "£39" or "£39.00")
 */
export function formatPrice(price: number, showPence: boolean = false): string {
    if (showPence || price % 1 !== 0) {
        return `£${price.toFixed(2)}`;
    }
    return `£${price}`;
}

/**
 * Calculate discount percentage for annual plans
 *
 * @param tier - Tier name
 * @returns Discount percentage (e.g., 20 for 20%)
 */
export function getAnnualDiscountPercentage(tier: PricingTier): number {
    const monthly = getTierPrice(tier, false);
    const annual = getTierPrice(tier, true);

    const fullYearCost = monthly * 12;
    const savings = fullYearCost - annual;

    return Math.round((savings / fullYearCost) * 100);
}

// ============================================================================
// NEW: EXPENSE TRACKING & REVENUE RECOVERY PRICING (V4)
// ============================================================================

/**
 * NEW PRICING STRUCTURE (Post-Expense Feature Launch)
 *
 * Focus: Revenue Recovery (Client Recharges + Tax Savings)
 * Positioning: "Find money you're leaving on the table"
 *
 * Tiers:
 * - Free: £0 - Basic expense tracking + invoicing
 * - Pro: £10/month - Unlimited expenses, OCR, advanced features
 * - MTD-Pro: £20/month - All Pro + HMRC quarterly filing
 */

export type ExpensePricingTier = 'free' | 'pro' | 'mtd-pro';

export interface ExpenseTierPricing {
    id: ExpensePricingTier;
    name: string;
    monthly: number;
    annual: number;
    annualSavings: number;

    // Expense limits
    expensesPerMonth: number | null; // null = unlimited
    receiptStorageMB: number;
    ocrProcessing: number | null; // null = unlimited

    // Invoice limits (keep existing functionality)
    invoicesPerMonth: number | null;
    collectionsPerMonth: number | null;
    collectionsChannels: string[];

    // Features
    features: string[];

    // Stripe Price IDs (TODO: Create in Stripe Dashboard)
    stripeMonthlyPriceId?: string;
    stripeYearlyPriceId?: string;
}

export const EXPENSE_PRICING_TIERS: Record<ExpensePricingTier, ExpenseTierPricing> = {
    free: {
        id: 'free',
        name: 'Free',
        monthly: 0,
        annual: 0,
        annualSavings: 0,

        // Expenses
        expensesPerMonth: 50,
        receiptStorageMB: 100,
        ocrProcessing: 10,

        // Invoices (limited)
        invoicesPerMonth: 10,
        collectionsPerMonth: 1,
        collectionsChannels: ['email'],

        features: [
            '50 expenses per month',
            '10 invoices per month',
            'Receipt upload & OCR (10/month)',
            'Billable expense tracking',
            'Revenue recovery dashboard',
            'Basic collections (1/month)',
            'Export data (CSV)',
        ],
    },

    pro: {
        id: 'pro',
        name: 'Pro',
        monthly: 10,
        annual: 96, // 20% discount = £8/month
        annualSavings: 24,

        // Expenses
        expensesPerMonth: null, // Unlimited
        receiptStorageMB: 1000, // 1GB
        ocrProcessing: null, // Unlimited

        // Invoices
        invoicesPerMonth: null, // Unlimited
        collectionsPerMonth: 25,
        collectionsChannels: ['email', 'sms'],

        features: [
            'Unlimited expenses & invoices',
            'Unlimited receipt OCR',
            '1GB receipt storage',
            'Advanced collections (25/month)',
            'Email + SMS reminders',
            'Bulk expense import',
            'AI revenue forecasting',
            'Branded invoices',
            'Client expense reports',
        ],

        stripeMonthlyPriceId: 'price_xxx', // TODO: Replace with real Stripe price IDs
        stripeYearlyPriceId: 'price_yyy',
    },

    'mtd-pro': {
        id: 'mtd-pro',
        name: 'MTD-Pro',
        monthly: 20,
        annual: 192, // 20% discount = £16/month
        annualSavings: 48,

        // All Pro features
        expensesPerMonth: null,
        receiptStorageMB: 1000,
        ocrProcessing: null,
        invoicesPerMonth: null,
        collectionsPerMonth: null, // Unlimited
        collectionsChannels: ['email', 'sms', 'phone'],

        features: [
            'All Pro features',
            'HMRC quarterly submissions',
            'VAT filing integration',
            'Annual tax declarations',
            'Audit-proof digital records',
            'Compliance reports',
            'Priority support',
            'Early access to new features',
        ],

        stripeMonthlyPriceId: 'price_zzz', // TODO: Replace with real Stripe price IDs
        stripeYearlyPriceId: 'price_aaa',
    },
};

/**
 * Get expense pricing for a tier
 */
export function getExpenseTierPrice(tier: ExpensePricingTier, isAnnual: boolean = false): number {
    const pricing = EXPENSE_PRICING_TIERS[tier];
    return isAnnual ? pricing.annual : pricing.monthly;
}

/**
 * Get expense limit for a tier
 */
export function getExpensesLimit(tier: ExpensePricingTier): number | null {
    return EXPENSE_PRICING_TIERS[tier].expensesPerMonth;
}

/**
 * Check if user has exceeded expense limit
 */
export function hasExceededExpenseLimit(
    tier: ExpensePricingTier,
    expensesUsed: number
): boolean {
    const limit = getExpensesLimit(tier);

    // Unlimited tiers never exceed
    if (limit === null) return false;

    return expensesUsed >= limit;
}

/**
 * Get recommended expense tier upgrade
 */
export function getRecommendedExpenseTierUpgrade(
    currentTier: ExpensePricingTier,
    expensesUsed: number
): ExpensePricingTier | null {
    // Already on top tier
    if (currentTier === 'mtd-pro') return null;

    const limit = getExpensesLimit(currentTier);

    // No limit = no upgrade needed
    if (limit === null) return null;

    // Recommend upgrade if using >80% of limit
    const usagePercentage = (expensesUsed / limit) * 100;

    if (usagePercentage >= 80) {
        if (currentTier === 'free') return 'pro';
        if (currentTier === 'pro') return 'mtd-pro';
    }

    return null;
}

/**
 * Format expense tier for display
 */
export function formatExpenseTier(tier: ExpensePricingTier): string {
    const tierMap: Record<ExpensePricingTier, string> = {
        free: 'Free',
        pro: 'Pro',
        'mtd-pro': 'MTD-Pro',
    };

    return tierMap[tier];
}
